File: cpp/lang_compare_client_tests.cc

#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>
#include "lang_compare.grpc.pb.h"

using namespace std;

using grpc::Channel;
using grpc::ClientContext;
using grpc::Status;

using langcompare::XorCipherRequest;
using langcompare::XorCipherReply;
using langcompare::LangCompare;

class LangCompareClient {
public:
    LangCompareClient(shared_ptr <Channel> channel)
            : stub_(LangCompare::NewStub(channel)) {}

    // Assembles the client's payload, sends it and presents the response back
    // from the server.
    string xorCipher(const string &key, const string &in_str) {
        // Data we are sending to the server.
        XorCipherRequest request;
        request.set_key(key);
        request.set_in_str(in_str);

        // Container for the data we expect from the server.
        XorCipherReply reply;

        // Context for the client. It could be used to convey extra information to
        // the server and/or tweak certain RPC behaviors.
        ClientContext context;

        // The actual RPC.
        Status status = stub_->XorCipher(&context, request, &reply);

        // Act upon its status.
        if (status.ok()) {
            return reply.out_str();
        } else {
            cout << status.error_code() << ": " << status.error_message() << endl;
            return "RPC failed";
        }
    }

private:
    unique_ptr <LangCompare::Stub> stub_;
};

void test_xor_cipher_cpp() {
    LangCompareClient compare(grpc::CreateChannel("localhost:50051", grpc::InsecureChannelCredentials()));
    int num = 1000;
    cout << "calling cpp xor_cipher " << num * 2 << " time" << endl;
    auto key = "my key";
    for (int i = 0; i < num - 1; ++i) {
        string in_str("message");
        in_str+= to_string(num);
        string encrypted_reply = compare.xorCipher(key, in_str);
        string decrypted_reply = compare.xorCipher(key, encrypted_reply);
    }
    string in_str("message");
    in_str+= to_string(num - 1);
    string encrypted_reply = compare.xorCipher(key, in_str);
    string decrypted_reply = compare.xorCipher(key, encrypted_reply);
    cout << "last message: " << decrypted_reply << endl;
}

int main(int argc, char **argv) {
    test_xor_cipher_cpp();

    return 0;
}

File: python/lang_compare_client_tests.py
import logging

import grpc

import lang_compare_pb2
import lang_compare_pb2_grpc


# noinspection DuplicatedCode
def test_xor_cipher_python():
    port_python = 'localhost:50052'
    print("Connecting to python lang-compare server at: {}".format(port_python))
    channel_python = grpc.insecure_channel(port_python)
    stub_python = lang_compare_pb2_grpc.LangCompareStub(channel_python)

    key = "my key"
    num = 1000
    print("calling python xor_cipher {} times".format(num * 2))
    decrypted = ""
    for i in range(num):
        in_str = "message{0:05d}".format(i)
        response_python = stub_python.XorCipher(lang_compare_pb2.XorCipherRequest(in_str=in_str, key=key))
        response_cpp = stub_python.XorCipher(lang_compare_pb2.XorCipherRequest(in_str=response_python.out_str, key=key))
        decrypted = response_cpp.out_str
        assert (decrypted == in_str)

    print("last decrypted string: '{}'\n".format(decrypted))


if __name__ == '__main__':
    logging.basicConfig()
    test_xor_cipher_python()
